policy/modules/contrib/frr.te

```diff
@@ -0,0 +1,108 @@
+policy_module(frr, 1.0.0)
+
+########################################
+#
+# Declarations
+#
+
+type frr_t;
+type frr_exec_t;
+init_daemon_domain(frr_t, frr_exec_t)
+
+type frr_log_t;
+logging_log_file(frr_log_t)
+
+type frr_tmp_t;
+files_tmp_file(frr_tmp_t)
+
+type frr_lock_t;
+files_lock_file(frr_lock_t)
+
+type frr_conf_t;
+files_config_file(frr_conf_t)
+
+type frr_unit_file_t;
+systemd_unit_file(frr_unit_file_t)
+
+type frr_var_run_t;
+files_pid_file(frr_var_run_t)
+
+########################################
+#
+# frr local policy
+#
+allow frr_t self:capability { dac_override dac_read_search chown kill net_bind_service net_raw setgid setuid };
+allow frr_t self:process { setcap setpgid };
+allow frr_t self:packet_socket create;
+allow frr_t self:netlink_route_socket rw_netlink_socket_perms;
+allow frr_t self:tcp_socket { connect connected_stream_socket_perms };
+allow frr_t self:udp_socket create_socket_perms;
+allow frr_t self:rawip_socket create_socket_perms;
+allow frr_t self:unix_stream_socket connectto;
+
+allow frr_t frr_conf_t:dir list_dir_perms;
+read_files_pattern(frr_t, frr_conf_t, frr_conf_t)
+read_lnk_files_pattern(frr_t, frr_conf_t, frr_conf_t)
+
+manage_dirs_pattern(frr_t, frr_log_t, frr_log_t)
+manage_files_pattern(frr_t, frr_log_t, frr_log_t)
+manage_lnk_files_pattern(frr_t, frr_log_t, frr_log_t)
+logging_log_filetrans(frr_t, frr_log_t, { dir file lnk_file })
+
+allow frr_t frr_tmp_t:file map;
+manage_dirs_pattern(frr_t, frr_tmp_t, frr_tmp_t)
+manage_files_pattern(frr_t, frr_tmp_t, frr_tmp_t)
+
+manage_files_pattern(frr_t, frr_lock_t, frr_lock_t)
+manage_lnk_files_pattern(frr_t, frr_lock_t, frr_lock_t)
+files_lock_filetrans(frr_t, frr_lock_t, { file lnk_file })
+
+manage_dirs_pattern(frr_t, frr_var_run_t, frr_var_run_t)
+manage_files_pattern(frr_t, frr_var_run_t, frr_var_run_t)
+manage_lnk_files_pattern(frr_t, frr_var_run_t, frr_var_run_t)
+manage_sock_files_pattern(frr_t, frr_var_run_t, frr_var_run_t)
+files_pid_filetrans(frr_t, frr_var_run_t, { dir file lnk_file })
+
+allow frr_t frr_exec_t:dir search_dir_perms;
+read_lnk_files_pattern(frr_t, frr_exec_t, frr_exec_t)
```

Comment:
Is there an example for executable symlink?

Reply:
Yes, here you have example. 

Example : `type=PROCTITLE msg=audit(02/02/2021 06:02:14.242:651) : proctitle=/usr/lib/frr/fabricd -d -A 127.0.0.1 
type=PATH msg=audit(02/02/2021 06:02:14.242:651) : item=0 name=/usr/lib64/frr/libfrr.so.0 nametype=UNKNOWN cap_fp=none cap_fi=none cap_fe=0 cap_fver=0 cap_frootid=0 
type=CWD msg=audit(02/02/2021 06:02:14.242:651) : cwd=/etc/frr 
type=SYSCALL msg=audit(02/02/2021 06:02:14.242:651) : arch=x86_64 syscall=openat success=no exit=EACCES(Permission denied) a0=0xffffff9c a1=0x7ffcae5eadb0 a2=O_RDONLY|O_CLOEXEC a3=0x0 items=1 ppid=35891 pid=35892 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=fabricd exe=/usr/lib/frr/fabricd subj=system_u:system_r:frr_t:s0 key=(null) 
type=AVC msg=audit(02/02/2021 06:02:14.242:651) : avc:  denied  { read } for  pid=35892 comm=fabricd name=libfrr.so.0 dev="vda1" ino=35651802 scontext=system_u:system_r:frr_t:s0 tcontext=system_u:object_r:frr_exec_t:s0 tclass=lnk_file permissive=0 
`

Reply:
Thanks, I managed to find the previous discussion.

Reply:
Can you post the output of this command?
```
ls -laZ /usr/lib64/frr
```

Reply:
This seems like a packaging bug in `frr` - the executable files currently in `/usr/lib/frr` should be placed in `/usr/libexec/frr` instead. Did you report it to the maintainer/upstream?

Reply:
Here:

- drwxr-xr-x.  3 root root system_u:object_r:frr_exec_t:s0     113 Feb  2 04:42 .
- dr-xr-xr-x. 49 root root system_u:object_r:lib_t:s0        24576 Feb  2 04:42 ..
- lrwxrwxrwx.  1 root root system_u:object_r:frr_exec_t:s0      19 Sep  3  2019 libfrrsnmp.so.0 -> libfrrsnmp.so.0.0.0
- -rwxr-xr-x.  1 root root system_u:object_r:frr_exec_t:s0   20792 Sep  3  2019 libfrrsnmp.so.0.0.0
- lrwxrwxrwx.  1 root root system_u:object_r:frr_exec_t:s0      15 Sep  3  2019 libfrr.so.0 -> libfrr.so.0.0.0
- -rwxr-xr-x.  1 root root system_u:object_r:frr_exec_t:s0 1200856 Sep  3  2019 libfrr.so.0.0.0
- drwxr-xr-x.  2 root root system_u:object_r:frr_exec_t:s0     151 Feb  2 04:42 modules

Reply:
This does not conform to the fc file. What do restorecon/matchpathcon say?
```
restorecon -Rvn /usr/lib64/frr
matchpathcon /usr/lib64/frr/libfrr.so.0 
```

Reply:
The problem is that in .fc there is `/usr/lib/frr(/.*)?              gen_context(system_u:object_r:frr_exec_t,s0)`, which breaks the labels in `/usr/lib64/frr/...`, since there is an [equivalency](https://src.fedoraproject.org/rpms/selinux-policy/blob/master/f/file_contexts.subs_dist#_8) between `/usr/lib` and `/usr/lib64`.

Reply:
@zpytela , matchpathcon say:
`/usr/lib64/frr/libfrr.so.0	system_u:object_r:frr_exec_t:s0`

Reply:
> This seems like a packaging bug in `frr` - the executable files currently in `/usr/lib/frr` should be placed in `/usr/libexec/frr` instead. Did you report it to the maintainer/upstream?

No didn't report that.

Reply:
> The problem is that in .fc there is `/usr/lib/frr(/.*)? gen_context(system_u:object_r:frr_exec_t,s0)`, which breaks the labels in `/usr/lib64/frr/...`, since there is an [equivalency](https://src.fedoraproject.org/rpms/selinux-policy/blob/master/f/file_contexts.subs_dist#_8) between `/usr/lib` and `/usr/lib64`.

So issue is that this files should be in `/usr/libexec/frr` a also have different label?

Reply:
Executable files which are not supposed to be executed from cmdline should be in /usr/libexec. Shared libraries should be in /usr/lib64 and have the common type lib_t unless there was a reason for having a private type - this matches the purpose of shared libraries.

Reply:
@WOnder93 So I have already reported this bug to frr maintainer.